AEROGEAR-8186 Move existing prometheus metrics functionality from syn…

aliok opened the pull request:
…c server to its own library

As we decided, I picked the simplest approach.

Changes in examples/metrics/server.js is done to demonstrate how to make use of these changes.

Verificaiton:
1. Start it by `node examples/metrics/server.js `
2. ~~Execute this query bunch of times~~
```
query hello {
  getUser{name, favoriteHorse{name, breed}}
}
```
UPDATE: the metrics example is simplified now. Just execute `query hello {hello}`.
3. Go to http://localhost:4000/metrics
You will see these:
```
# HELP resolver_timing_ms Resolver response time in milliseconds
# TYPE resolver_timing_ms histogram
resolver_timing_ms_bucket{le="0.005",operation_type="query",name="Query.hello"} 0
resolver_timing_ms_bucket{le="0.01",operation_type="query",name="Query.hello"} 0
resolver_timing_ms_bucket{le="0.025",operation_type="query",name="Query.hello"} 0
resolver_timing_ms_bucket{le="0.05",operation_type="query",name="Query.hello"} 0
resolver_timing_ms_bucket{le="0.1",operation_type="query",name="Query.hello"} 0
resolver_timing_ms_bucket{le="0.25",operation_type="query",name="Query.hello"} 0
resolver_timing_ms_bucket{le="0.5",operation_type="query",name="Query.hello"} 0
resolver_timing_ms_bucket{le="1",operation_type="query",name="Query.hello"} 1
resolver_timing_ms_bucket{le="2.5",operation_type="query",name="Query.hello"} 1
resolver_timing_ms_bucket{le="5",operation_type="query",name="Query.hello"} 1
resolver_timing_ms_bucket{le="10",operation_type="query",name="Query.hello"} 1
resolver_timing_ms_bucket{le="+Inf",operation_type="query",name="Query.hello"} 1
resolver_timing_ms_sum{operation_type="query",name="Query.hello"} 1
resolver_timing_ms_count{operation_type="query",name="Query.hello"} 1

# HELP requests_resolved Number of requests resolved by server
# TYPE requests_resolved counter
requests_resolved{operation_type="query",path="Query.hello"} 1

# HELP requests_resolved_total Number of requests resolved by server in total
# TYPE requests_resolved_total counter
requests_resolved_total{operation_type="query",path="Query.hello"} 1

# HELP server_response_ms Server response time in milliseconds
# TYPE server_response_ms histogram
server_response_ms_bucket{le="0.005",request_type="GET",error="false"} 0
server_response_ms_bucket{le="0.01",request_type="GET",error="false"} 0
server_response_ms_bucket{le="0.025",request_type="GET",error="false"} 0
server_response_ms_bucket{le="0.05",request_type="GET",error="false"} 0
server_response_ms_bucket{le="0.1",request_type="GET",error="false"} 0
server_response_ms_bucket{le="0.25",request_type="GET",error="false"} 0
server_response_ms_bucket{le="0.5",request_type="GET",error="false"} 0
server_response_ms_bucket{le="1",request_type="GET",error="false"} 0
server_response_ms_bucket{le="2.5",request_type="GET",error="false"} 0
server_response_ms_bucket{le="5",request_type="GET",error="false"} 1
server_response_ms_bucket{le="10",request_type="GET",error="false"} 2
server_response_ms_bucket{le="+Inf",request_type="GET",error="false"} 2
server_response_ms_sum{request_type="GET",error="false"} 10
server_response_ms_count{request_type="GET",error="false"} 2
server_response_ms_bucket{le="0.005",request_type="GET",error="true"} 0
server_response_ms_bucket{le="0.01",request_type="GET",error="true"} 0
server_response_ms_bucket{le="0.025",request_type="GET",error="true"} 0
server_response_ms_bucket{le="0.05",request_type="GET",error="true"} 0
server_response_ms_bucket{le="0.1",request_type="GET",error="true"} 0
server_response_ms_bucket{le="0.25",request_type="GET",error="true"} 0
server_response_ms_bucket{le="0.5",request_type="GET",error="true"} 0
server_response_ms_bucket{le="1",request_type="GET",error="true"} 0
server_response_ms_bucket{le="2.5",request_type="GET",error="true"} 1
server_response_ms_bucket{le="5",request_type="GET",error="true"} 1
server_response_ms_bucket{le="10",request_type="GET",error="true"} 1
server_response_ms_bucket{le="+Inf",request_type="GET",error="true"} 1
server_response_ms_sum{request_type="GET",error="true"} 2
server_response_ms_count{request_type="GET",error="true"} 1
server_response_ms_bucket{le="0.005",request_type="POST",error="false"} 0
server_response_ms_bucket{le="0.01",request_type="POST",error="false"} 0
server_response_ms_bucket{le="0.025",request_type="POST",error="false"} 0
server_response_ms_bucket{le="0.05",request_type="POST",error="false"} 0
server_response_ms_bucket{le="0.1",request_type="POST",error="false"} 0
server_response_ms_bucket{le="0.25",request_type="POST",error="false"} 0
server_response_ms_bucket{le="0.5",request_type="POST",error="false"} 0
server_response_ms_bucket{le="1",request_type="POST",error="false"} 0
server_response_ms_bucket{le="2.5",request_type="POST",error="false"} 0
server_response_ms_bucket{le="5",request_type="POST",error="false"} 1
server_response_ms_bucket{le="10",request_type="POST",error="false"} 1
server_response_ms_bucket{le="+Inf",request_type="POST",error="false"} 2
server_response_ms_sum{request_type="POST",error="false"} 38
server_response_ms_count{request_type="POST",error="false"} 2```

darahayes commented on packages/apollo-voyager-metrics/src/index.ts:
```diff
@@ -33,64 +31,138 @@ const serverResponseMetric = new Prometheus.Histogram({
-  res.on('finish', onResFinished)
-  res.on('error', onResFinished)
+export function applyResponseLoggingMetricsMiddleware (app: Application) {
+  app.use(responseLoggingMetric as (req: IncomingMessage, res: Response, next: NextFunction) => void)
```
Eventually we might be able to turn this into an interface and support more http frameworks than just express so +1 for this.

darahayes commented on packages/apollo-voyager-metrics/src/index.ts:
```diff
@@ -33,64 +31,138 @@ const serverResponseMetric = new Prometheus.Histogram({
 
-  if (next) { next() }
+export function applyMetricsMiddleware (app: Application) {
+  app.get('/metrics', getMetrics)
```
Same as above. +1  for the idea but I feel like the end user should be able to configure the endpoint. Is there any strong reason why they shouldn't be able to, in terms of APBs or things like that?

aliok replied:
I did it that way because I saw `applyMiddleware` in other places. Wanted to keep it consistent. No strong opinion though. Can do it that way.

darahayes replied:
Sounds good. Just for context, the `applyMiddleware` pattern comes from the various apollo server integrations such as `apollo-server-express`, `apollo-server-hapi` `apollo-server-koa`. <- those libs are actually very thin wrappers that extend the base `ApolloServer` class with a framework specific `applyMiddleware` function that knows how to do the correct setup for a given framework.

We're not passing in any options into `server.applyMiddleware()` in the example apps but it does accept an options object with some things like a path. We can come back to it later but definitely we'll want that at some stage. We may eventually want to go down a similar route where we can support more than just express. 🙂

darahayes commented on packages/apollo-voyager-server/package.json:
```diff
@@ -24,6 +24,7 @@
   },
   "dependencies": {
     "@aerogear/apollo-voyager-context": "^1.0.0",
+    "@aerogear/apollo-voyager-metrics": "^1.0.0",
```
I don't see any changes in the apollo-voyager-server package. Is there any reason why we need this here?

darahayes commented on packages/apollo-voyager-metrics/src/index.ts:
```diff
@@ -33,64 +31,138 @@ const serverResponseMetric = new Prometheus.Histogram({
-  const requestMethod = req.method as string
+export function wrapResolversForMetrics (resolvers: {[key: string]: ResolverObject}): {[key: string]: ResolverObject} {
+  const output: {[key: string]: ResolverObject} = {}
+  for (const typeKey in resolvers) {
```
I find this code slightly hard to understand. Between the typescript types and the `for ... in` and `continue` statements, I find this quite syntax heavy. 

I think maybe it might be worth renaming `resolvers` to `resolverMappings`

And then doing something like this.

```js
const typeKeys = Object.keys(resolverMappings)

typeKeys.forEach((typeKey) => {
  const type = resolverMappings[typeKey]
  const resolversKeysForType = Object.keys(type)
  result[typeKey] = {}

  resolversKeysForType.forEach((resolverKey) => {
    const resolverFn = type[resolverKey]
    output[typeKey][resolverKey] = wrapSingleResolverForMetrics(resolverFn)
  })
})

```

To be honest... I don't really know if that even improves it. Feel free to ignore it :)

aliok replied:
Ok, I am +1 on using `Object.keys()` so that `hasOwnProperty` check is not necessary. That would eliminate the `continue` statements.
But I like the `for in` loop. IMO more straightforward than the `.forEach` thing.

aliok commented:
@darahayes thx for the review.
I simplified the metrics example. But make sure you also do verification using nested resolvers.

Applied changes that addresses some of the comments you made. I left some unaddressed.

darahayes commented:
@aliok looks great. I will verify soon!
